### PR TITLE
Move constructor for Triangulation

### PR DESCRIPTION
**Pull request by danshapero**
This PR also changes some of the behavior in the destructor of Triangulation. Any pointer members are checked before deleting, and set to 0 after. I've run all the triangulation tests; since this is change is pretty far-reaching, I can run more of the test suite if need be.

### CODE REVIEW

File: doc/news/changes.h [hunk 1]
```diff
@@ -336,6 +336,11 @@ inconvenience this causes.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> New: a move constructor has been added to Triangulation.
```
Comment (bangerth) [inline doc/news/changes.h]: a -> A

File: source/grid/tria.cc [hunk 2]
```diff
@@ -8995,28 +8995,79 @@ Triangulation (const Triangulation<dim, spacedim> &other)
 
 
 
+#ifdef DEAL_II_WITH_CXX11
+template <int dim, int spacedim>
+Triangulation<dim, spacedim>::
+Triangulation (Triangulation<dim, spacedim> &&tria)
+  :
+  Subscriptor(static_cast<Subscriptor&&>(tria)),
```
Comment (bangerth) [inline source/grid/tria.cc]: do you actually need the `static_cast` here?

File: source/grid/tria.cc [hunk 3]
```diff
@@ -8995,28 +8995,79 @@ Triangulation (const Triangulation<dim, spacedim> &other)
...
+  smooth_grid(tria.smooth_grid),
+  periodic_face_pairs_level_0(tria.periodic_face_pairs_level_0),
+  periodic_face_map(tria.periodic_face_map),
+  levels(tria.levels),
+  faces(tria.faces),
+  vertices(std::move(tria.vertices)),
+  vertices_used(tria.vertices_used),
```
Comment (bangerth) [inline source/grid/tria.cc]: out of curiosity, why is it that you need the `std::move` in the first of these two lines, but not in the second?

**Comment by bangerth** [comment]
Nice -- at some future time, we may use this to convert all of the `GridGenerator` classes to actually *return* the triangulation, rather than passing it back as an argument. (Though that may also require a move assignment operator.)

File: source/grid/tria.cc (hunk 3 above)
Comment (danshapero) [inline source/grid/tria.cc]: I wrote this a few weeks ago and forgot why I had that. Without it, the check that the moved triangulation is empty afterwards fails because it still has vertices. I read up on this some more today; apparently the `std::move` is necessary, because it converts an r-value reference to an x-value. This makes me think that all the other members that are of type `std::vector` should be moved as well. I'll experiment with this and try it on periodic triangulations or triangulations with manifold boundaries to be sure that it works right in all other cases. Could you relabel it as work-in-progress for now?

Comment (bangerth) [inline source/grid/tria.cc]: Sure. Thanks for experimenting. I would have expected that `rvalue_reference.member_variable` also ends up being an rvalue reference, so that this calls the move constructor of the member variable to be initialized, but I may of course be wrong about this. I often end up writing little 20-line programs to test this when I get confused.

Comment (danshapero) [inline source/grid/tria.cc]: I think [this SO answer](http://stackoverflow.com/a/14486705/1762306) should clarify. It's very confusing.

Comment (bangerth) [inline source/grid/tria.cc]: Ah, yes, good find. The second answer is what I was needing. I guess that means that in member initializer lists, one should generally use `std::move` for everything that's not a scalar. Can you adjust your patch accordingly?

(Incidentally, I think we should also look at the move constructors for other classes that have recently obtained move constructors.)

**Comment by danshapero** [comment]
@bangerth -- I added more to the patch, but I want to test it more when the triangulation is periodic.

@tjhei and I were just talking about move constructors. Part of the reason I suggested in #2490 that a moved object be empty is that I can check for emptiness in the unit test to be sure that copy semantics hasn't been used inadvertently. This would be easier for Triangulation if had a method to check for emptiness, which I could write as part of this patch if you think that would be better. I'll go over the move constructors I've already written again, but for the most part they check that the former husk of the moved object is empty.

This brings up a more interesting problem. Classes like Quadrature which don't manage heap memory can use the default move constructor, while for classes like Vector or SparseMatrix that do manage memory, it's unlikely that anyone will be adding new members (right?). Triangulation has the worst of both cases; periodicity and manifolds have been added just in the past year, but it contains raw pointers. If we could use `unique_ptr` or an index into some other data structure, Triangulation could instead use the default move constructor.

**Comment by bangerth** [comment]
@danshapero -- great, and yes, more tests are more better. Feel free to add multiple tests that are slight variations of each other to test different possible corner cases. We often have whole sequences of tests that are all just variations of each other, and I think that's just fine.

Re `empty()` -- yes, if that makes your life easier, add it. In fact, many functions in namespace `GridGenerator` require that the triangulation they fill be empty when it is passed in, and they currently check this in awkward or tangential ways, of at all. It would be easier if there was `Triangulation::empty`.

As for pointers: yes, Triangulation is an old class. It was written before shared or unique pointers existed. In more recent classes, we have consistently used `std_cxx11::unique_ptr`, and I certainly would think that making such a conversion would be a useful step in itself. I would probably put that into a separate pull request, but that may create additional work and so I'll also be ok if you wanted to do it as part of the current pull request.

**Comment by bangerth** [comment]
In reference to #898.

**Comment by bangerth** [comment]
@danshapero -- let us know once you are ready and want a re-review!

**Comment by danshapero** [comment]
Will do! I've been out of the office for the past few weeks. I need to add one more test and I'll fix the merge conflicts.

**Comment by danshapero** [comment]
Alright, I've written a test for moving triangulations with periodicity and I'm ready for another review.

File: source/grid/tria.cc [hunk 4]
```diff
@@ -8996,28 +8996,80 @@ Triangulation (const Triangulation<dim, spacedim> &other)
...
+
 template <int dim, int spacedim>
 Triangulation<dim, spacedim>::~Triangulation ()
 {
   for (unsigned int i=0; i<levels.size(); ++i)
-    delete levels[i];
+    if (levels[i])
+      {
+        delete levels[i];
+        levels[i] = 0;
```
Comment (bangerth) [inline source/grid/tria.cc]: yes, nice safety measure

**Comment by bangerth** [comment]
Fantastic. Ready to merge once the tester is happy.

Merging this PR will also enable #2934 and #2935.

**Comment by bangerth** [comment]
Oh, no, I merged another PR before yours and that created a merge conflict in the `changes.h` file. Could you please rebase on current master and fix the conflict?

**Comment by danshapero** [comment]
Sure thing, should be fixed now. I'll rebase down to a sane commit history if you think it's good to go.

**Comment by bangerth** [comment]
Yes, ready to go. Please squash some of the commits (I usually go down to only one commit, or one each for code changes/changelog/testcases).

File: source/grid/tria.cc (hunk 4 above)
Comment (mikeando) [inline source/grid/tria.cc]: Though of course the `if` is unneeded right? `delete` on a `nullptr` is fine.

File: source/grid/tria.cc [hunk 5]
```diff
@@ -8996,28 +8996,80 @@ Triangulation (const Triangulation<dim, spacedim> &other)
...
+  for (unsigned int i=0; i<tria.levels.size(); ++i)
+    tria.levels[i] = nullptr;
+
+  tria.faces = nullptr;
+
+  tria.number_cache = internal::Triangulation::NumberCache<dim>();
+
+  tria.vertex_to_boundary_id_map_1d = nullptr;
+  tria.vertex_to_manifold_id_map_1d = nullptr;
+}
```
Comment (mikeando) [inline source/grid/tria.cc]: Probably worth pulling the bulk of the invalidation of the input into a function so it can be reused when a move assignment operator is added.

Adding a move constructor to `NumberCache<dim>` would also make this a little nicer.

Comment (bangerth) [inline source/grid/tria.cc]: Most of the invalidation can disappear again if we make these variables `std::unique_ptr` objects as the move constructor for them will already do the invalidation. See #2934.

File: source/grid/tria.cc (hunk 4 above)
Comment (bangerth) [inline source/grid/tria.cc]: Probably, but also doesn't hurt. As before, this should be a moot point with #2934.

**Comment by bangerth** [comment]
Darn, we can't win. There is again a merge conflict :-( Can you rebase one more time?

(We really need to come up with a better system to deal with our changelog file...)

**Comment by danshapero** [comment]
[Every time I rebase.](https://i.miximum.fr/i/2014/12/1X9YPBFMK8_l.png) I'm working on it.

**Comment by bangerth** [comment]
Thanks!
